pkg/deviceplugins/usb_device_plugin.go

```diff
@@ -0,0 +1,452 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2024 SUSE, LLC.
+ *
+ */
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	v1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/client-go/log"
+	"kubevirt.io/kubevirt/pkg/safepath"
+	"kubevirt.io/kubevirt/pkg/util"
+	pluginapi "kubevirt.io/kubevirt/pkg/virt-handler/device-manager/deviceplugin/v1beta1"
+
+	"github.com/harvester/pcidevices/pkg/apis/devices.harvesterhci.io/v1beta1"
+)
+
+var (
+	pathToUSBDevices = "/sys/bus/usb/devices"
+)
+
+// The actual plugin
+type USBDevicePlugin struct {
+	socketPath   string
+	stop         <-chan struct{}
+	update       chan struct{}
+	done         chan struct{}
+	deregistered chan struct{}
+	server       *grpc.Server
+	resourceName string
+	device       *PluginDevice
+	logger       *log.FilteredLogger
+
+	initialized bool
+	lock        *sync.Mutex
+}
+
+type PluginDevice struct {
+	ID           string
+	isHealthy    bool
+	DevicePath   string
+	Bus          int
+	DeviceNumber int
+}
+
+type USBDevicePluginInterface interface {
+	Start(stop <-chan struct{}) error
+}
+
+func (pd *PluginDevice) toKubeVirtDevicePlugin() *pluginapi.Device {
+	healthStr := pluginapi.Healthy
+	if !pd.isHealthy {
+		healthStr = pluginapi.Unhealthy
+	}
+	return &pluginapi.Device{
+		ID:       pd.ID,
+		Health:   healthStr,
+		Topology: nil,
+	}
+}
+
+func (plugin *USBDevicePlugin) GetDevice() *PluginDevice {
+	return plugin.device
+}
+
+func (plugin *USBDevicePlugin) setDeviceHealth(isHealthy bool) {
+	pd := plugin.GetDevice()
+	isDifferent := pd.isHealthy != isHealthy
+	pd.isHealthy = isHealthy
+	if isDifferent {
+		plugin.update <- struct{}{}
+	}
+}
+
+func (plugin *USBDevicePlugin) devicesToKubeVirtDevicePlugin() []*pluginapi.Device {
+	return []*pluginapi.Device{plugin.device.toKubeVirtDevicePlugin()}
+}
+
+func (plugin *USBDevicePlugin) GetInitialized() bool {
+	plugin.lock.Lock()
+	defer plugin.lock.Unlock()
+	return plugin.initialized
+}
+
+func (plugin *USBDevicePlugin) setInitialized(initialized bool) {
+	plugin.lock.Lock()
+	plugin.initialized = initialized
+	plugin.lock.Unlock()
+}
+
+func (plugin *USBDevicePlugin) GetDeviceName() string {
+	return plugin.resourceName
+}
+
+func (plugin *USBDevicePlugin) StopDevicePlugin() error {
+	defer func() {
+		select {
+		case <-plugin.done:
+			return
+		default:
+			close(plugin.done)
+		}
+	}()
+
+	// Give the device plugin one second to properly deregister
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+	select {
+	case <-plugin.deregistered:
+	case <-ticker.C:
+	}
```

Comment:
Just wait one second for the notification from `ListAndWatch()`, wouldn't it be too short?

Reply:
VGPU uses 5 second
https://github.com/Yu-Jack/pcidevices/blob/f22589e6ad959cad0d70987395c747b6833c7c5b/pkg/deviceplugins/vgpu_device_manager.go#L365-L370

@ibrokethecloud WDYT?

Reply:
After discuss with @ibrokethecloud , we all agree with 5 seconds.